lib/auth/default_current_user_provider.rb

```diff
@@ -240,14 +254,15 @@ def log_on_user(user, session, cookies, opts = {})
   end
 
   def cookie_hash(unhashed_auth_token, user)
-    cookie = DiscourseAuthCookie.new(
+    data = {
       token: unhashed_auth_token,
       user_id: user.id,
       trust_level: user.trust_level,
-      issued_at: Time.zone.now
-    )
-    hash = {
-      value: cookie.serialize,
+      issued_at: Time.zone.now.to_i
+    }
+    temp_request = ActionDispatch::Request.new(@env)
+    hash = temp_request.cookie_jar.encrypted[TOKEN_COOKIE] = {
```

Comment:
As I understand it, this line will immediately set the cookie (and persist it to `@env`, ready to be sent back to the user when the response is rendered).

Therefore, I'm not sure that having a `#cookie_hash` method makes sense any more. Instead, I think it should be more like `#set_token_cookie!(unhashed_auth_token, user)`, and then we should remove all the `_t` cookie setting from other places?

Reply:
So I've tried to make this change, however there are a couple of complications that came up:

1) the interface of `Auth::CurrentUserProvider` and `Auth::DefaultCurrentUserProvider`, specifically the signature `log_on_user` and `refresh_session` methods currently expect a `cookies` parameter which is a hash that the methods mutate to add/remove the auth cookie. If we change how the cookie is set so it's set on `@env` directly then the `cookies` parameter would no longer be used and we'd have to break the methods signature. From a quick search I can see a few plugins that'd have to be updated.

2) a couple of tests in `default_current_user_provider_spec.rb` needs to access the `expires`, `httponly` and `secure` attribute of the auth cookie, and they currently do that by passing a hash to the `log_on_user` and/or `refresh_session` methods which mutate the hash and then the tests access the cookie from the hash and do assertions. If we remove the `cookies` parameter, the only way to read/access the auth cookie is via a `CookieJar` instance in the `env` object. However, `CookieJar` doesn't have public APIs to read a cookie's attributes (there is an API to access the cookie value, but none for the httponly, secure etc. attributes of the cookie). We could do something like this `cookie_jar.instance_variable_get(:@set_cookies)["_t"][:expires]` but I believe that's reaching into a private rails API and I'm not sure we want that.

So there are tradeoffs here between cleaner/more optimized code vs introducing breaking changes and reaching for private rails API in tests. I'm leaning towards not changing APIs now for now because this change has gotten too big/complicated I'd prefer not to add extra complexity by adding plugins to the mix unless there are serious problems with the current implementation, but I'm curious what you and @tgxworld think?

I've pushed my (unfinished) changes so you can see what I'm talking about.

Reply:
Ah I see, this is very tricky... I would hate to involve plugin changes here!

Unfortunately we can't leave it how it was before your most recent commit, because `temp_request.cookie_jar.encrypted[TOKEN_COOKIE] = ...`  will **actually set the cookie in the user's browser**, not just return a hash. So I think we have a choice of either going back to handling the encryption/decryption ourselves, or trying to press forward.

For (1), the `cookies` parameter that we pass to the `log_on_user` and `refresh_session` is already an `ActionDispatch::Cookies::CookieJar` (we call these methods from Rails, where `cookies` resolves to the cookie jar). It's only the tests which pass simple hashes into these methods. So once we fix the tests, I think we could probably make this change without any method signature changes.

For (2), it is indeed unfortunate that the CookieJar doesn't provide any public APIs for inspecting the metadata of a cookie 😢. The way that we were previously using a hash in tests is kinda bad - the hashes were essentially poor 'mocks' for `ActionDispatch::Cookies::CookieJar` objects. So I think updating the specs to use a proper CookieJar is a good change. I think we might be able to avoid reaching into private instance variables by using the `#write` method, and looking at the header it produces? In a sense, this is actually a better 'end-to-end' test of the system.

I had a go at sketching out those two thoughts here: https://github.com/discourse/discourse/compare/feature/rate-limits-per-user...feature/rate-limits-per-user-david. Feel free to merge/discard any of those changes as you wish.

Reply:
This looks good to me, thank you David! Will apply these changes.

I'm just curious about this:

> Unfortunately we can't leave it how it was before your most recent commit, because `temp_request.cookie_jar.encrypted[TOKEN_COOKIE] = ...` will **actually set the cookie in the user's browser**, not just return a hash.

Isn't setting the cookie our end goal? the difference is we're doing double the work because cookie is set twice; the first time is when `temp_request.cookie_jar.encrypted[TOKEN_COOKIE] = ...` is executed and the second time when the `cookie_hash` method returns the hash and the same cookie is set again with `cookies[TOKEN_COOKIE] = cookie_hash(...)`. The problem with this of course is that the behavior of `cookie_hash` becomes a little surprising because it actually sets a cookie.

Reply:
This is done!

Reply:
> the difference is we're doing double the work because cookie is set twice... The problem with this of course is that the behavior of cookie_hash becomes a little surprising because it actually sets a cookie.

Yeah this was my concern - even though it may not mattter in the end, the seemingly unintended side effect of `cookie_hash` could be pretty surprising.